### PR TITLE
PWM fan support

### PR DESCRIPTION
**Pull request by pkendall64**
Adds support for PWM fan with a 'tachometer' pin as well.

Currently the PWM/speed of the fan is hard-coded based on the power output of the module and the reading of the tachometer is just printed as debug messages.

In future we could detect the speed of the fan and if it's not within  some range of expected speed, then downgrade the power and have a warning on the module (in LUA, OLED, RGB etc).

### CODE REVIEW

File: src/lib/THERMAL/devThermal.cpp [hunk 1]
```diff
@@ -110,11 +172,63 @@ static void timeoutFan()
...
+}
+
+static void timeoutTacho()
+{
+#if defined(PLATFORM_ESP32)
+    uint16_t pulses = tachoPulses;
+    tachoPulses = 0;
+
+    currentRPM = pulses * (60000 / THERMAL_DURATION) / TACHO_PULSES_PER_REV;
+    DBGLN("RPM %d", currentRPM);
```
Comment (CapnBry) [inline src/lib/THERMAL/devThermal.cpp:205]: DBGLN -> DBGVLN for things that spam

File: src/lib/THERMAL/devThermal.cpp [hunk 2]
```diff
@@ -110,11 +172,63 @@ static void timeoutFan()
...
-            digitalWrite(GPIO_PIN_FAN_EN, HIGH);
+            if (GPIO_PIN_FAN_EN != UNDEF_PIN)
+            {
+                digitalWrite(GPIO_PIN_FAN_EN, HIGH);
+            }
+#if defined(PLATFORM_ESP32)
+            else if (GPIO_PIN_FAN_PWM != UNDEF_PIN)
+            {
+                ledcWrite(fanChannel, fanSpeeds[POWERMGNT::currPower()]);
+                DBGLN("Fan speed: %d (power) -> %d (pwm)", POWERMGNT::currPower(), fanSpeeds[POWERMGNT::currPower()]);
```
Comment (CapnBry) [inline src/lib/THERMAL/devThermal.cpp:183]: Can you make this body into a function and call it from here and the identical code in the isOn block?

File: src/lib/THERMAL/devThermal.cpp [hunk 3]
```diff
@@ -86,7 +117,29 @@ static void timeoutFan()
...
+                    ledcWrite(fanChannel, fanSpeeds[POWERMGNT::currPower()]);
+                    DBGLN("Fan speed: %d (power) -> %d (pwm)", POWERMGNT::currPower(), fanSpeeds[POWERMGNT::currPower()]);
+                    lastPower = POWERMGNT::currPower();
+                    fanStateDuration = 0; // reset the timeout
+                }
+            }
+            else
+            {
+                fanStateDuration = 0; // reset the timeout
+            }
```
Comment (CapnBry) [inline src/lib/THERMAL/devThermal.cpp:139]: bump this block of code to outside the ifdef since it is the same I think?

File: src/lib/THERMAL/devThermal.cpp [hunk 4]
```diff
@@ -55,7 +85,7 @@ static void timeoutThermal()
     if(OPT_HAS_THERMAL_LM75A && !CRSF::IsArmed() && connectionState != wifiUpdate)
```
Comment (CapnBry) [inline src/lib/THERMAL/devThermal.cpp:85]: OFF TOPIC - Can anyone tell me what this is supposed to do with IsArmed? I'm not sure how the LM75 thresholds interact with the fan and is it possible we can remove this IsArmed check easily?

File: src/lib/THERMAL/devThermal.cpp [hunk 5]
```diff
@@ -35,14 +38,41 @@ bool is_smart_fan_working = false;
 
 #include "POWERMGNT.h"
 
+#if defined(GPIO_PIN_FAN_PWM)
+uint8_t fanSpeeds[] = {
```
Comment (CapnBry) [inline src/lib/THERMAL/devThermal.cpp:42]: static const

**Review by CapnBry — COMMENTED** [review]
Awesome, I was going to complain that there should be a way to turn it off but looks like you're using the fan thresh for this too, perfect! Just a couple minor style quibbles.

So this should just work on existing fans right? The PWM part not the tach part I mean

File: src/lib/THERMAL/devThermal.cpp (hunk 3 above)
Comment (pkendall64) [reply src/lib/THERMAL/devThermal.cpp:139]: I could if I early exited the function at the line above where it's incremented.

File: src/lib/THERMAL/devThermal.cpp (hunk 4 above)
Comment (pkendall64) [reply src/lib/THERMAL/devThermal.cpp:85]: I'm not sure why it wouldn't to thermal checking when armed or in wifi mode!? Colour me confuzzled 🤷‍♂️ 
This came from the original Axis PR, so I'll check on the Axis module what it's doing.

File: src/lib/THERMAL/devThermal.cpp (hunk 3 above)
Comment (CapnBry) [reply src/lib/THERMAL/devThermal.cpp:146]: I was thinkin' like this would support both ESP32 / everything else. It just makes it slightly easier to read for me, so if you're not down with that, you can just mark this as resolved. Everything I tagged was so minor 😁
```cpp
#if defined(PLATFORM_ESP32)
            if (GPIO_PIN_FAN_PWM != UNDEF_PIN)
            {
...
            }
            else
#else
            {
                fanStateDuration = 0; // reset the timeout
            }
#endif
```

File: src/lib/THERMAL/devThermal.cpp (hunk 4 above)
Comment (pkendall64) [reply src/lib/THERMAL/devThermal.cpp:85]: I've tracked it back to the original PR before we had the device framework, where it would have been running on the main thread along with everything else and was causing problems with the SPI for the radio. Back then it was also using software SPI for the TFT so things were very slow!

TL;DR I believe these checks can be removed.

**Comment by CapnBry** [comment]
Which hardware is this for specifically? I tested PWM mode with:
* Namimno OLED - Only works at full power (255 PWM)
* Thor Axis - Has auto fan control haha, so can't work
* HappyModel ProSeries - Works but can can only turn on at 95 PWM (100mW) but will run down to 63 PWM (50mW)

Fixes 
* The comment for `timeoutFan()` needs to be updated to two sections, one for PWM one for fixed
* Should `timeoutTacho()` have a `if (GPIO_PIN_FAN_TACHO == UNDEF_PIN) return;`? Also the RPM is unsigned in the log message %u
* `fanSpeeds` no longer needs to be in the unit scope, move it into setFanSpeed as just a const? pwm is also unsigned in the log message
* In `POWERMGNT::currPower() < lastPower && fanStateDuration >= FAN_MIN_CHANGETIME)`, the currPower check is redundant. The only way fanStateDuration can exceed the limit is if the power is less than lastPower. Removing this makes the next suggestion easy to implement.
* Because fans may have a hard time spinning up at the lowest possible speed, when the fan is turned on from being off, it should be bumped to a higher speed for one `THERMAL_DURATION` to be sure it turns on, before lowering to the proper speed. If the curPower check is removed from the line as suggested as above, this can be done by changing the fanShouldBeOn section to
```cpp
    // vv else fan is off currently vv
    else if (fanShouldBeOn)
    {
        // Delay turning the fan on for 4 cycles to be sure it really should be on
        if (fanStateDuration < 3)
        {
            ++fanStateDuration; // counting to turn on
        }
        else
        {
            if (GPIO_PIN_FAN_EN != UNDEF_PIN)
            {
                digitalWrite(GPIO_PIN_FAN_EN, HIGH);
                fanStateDuration = 0;
            }
#if defined(PLATFORM_ESP32)
            else if (GPIO_PIN_FAN_PWM != UNDEF_PIN)
            {
                // bump the fan to full power for one cycle in case
                // the PWM level is not sufficient to get it moving
                ledcWrite(fanChannel, 192);
                fanStateDuration = FAN_MIN_CHANGETIME;
            }
#endif
            fanIsOn = true;
        }
    }
#endif```

**Review by CapnBry — COMMENTED** [review]
This wouldn't be needed if you had removed the `POWERMGNT::currPower() < lastPower &&` as suggested instead of removing it from the `++fanStateDuration` test.

Now instead of waiting `FAN_MIN_CHANGETIME` _after_ the power goes under the correct level it will always lower the fan speed on the first tick that the power is lowered. The original code kept the fan running at the higher speed `FAN_MIN_CHANGETIME` after the power lowered. I thought that was the intent.

This new code has the possibility of incrementing the fanStateDuration until it rolls over if the power level doesn't go down, which then blocks the fan speed from changing until fanStateDuration goes above the threshold again. 😵

**Comment by pkendall64** [comment]
Sorry about that, I completely misread which line I was fixing! All fixed and tested now.

**Review by CapnBry — APPROVED** [review]
Works great! And I can hear the fan kick on at a higher power then throttle back a second later. Seems perfect now. Whew what a journey, sorry to drag you along on it 😄

**Comment by pkendall64** [comment]
That's what reviews are for 😄

File: src/lib/THERMAL/rpm.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#include "logging.h"
+#include <driver/pcnt.h>
+#include <soc/pcnt_struct.h>
+
+static volatile int overflow = 0;
+static uint32_t lastTime = 0;
+
+static void IRAM_ATTR pcnt_overflow(void *arg)
+{
+    overflow++;
```
Comment (SunjunKim) [inline src/lib/THERMAL/rpm.cpp:13]: What happens when `overflow` overflows?
Doesn't it need to reset `overflow=0` somewhere?

File: src/lib/THERMAL/rpm.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    pcnt_counter_resume(PCNT_UNIT_0);
+}
+
+uint32_t get_rpm()
+{
+    int16_t counter;
+    pcnt_get_counter_value(PCNT_UNIT_0, &counter);
+    pcnt_counter_clear(PCNT_UNIT_0);
+    uint32_t now = millis();
+    uint32_t rpm = ((overflow * 30000) + counter) * 60000 / 4 / (now - lastTime);
```
Comment (SunjunKim) [inline src/lib/THERMAL/rpm.cpp:54]: `overflow=0` after this?

File: src/lib/THERMAL/rpm.cpp (hunk 6 above)
Comment (pkendall64) [reply src/lib/THERMAL/rpm.cpp:13]: Yup

File: src/lib/THERMAL/rpm.cpp (hunk 7 above)
Comment (pkendall64) [reply src/lib/THERMAL/rpm.cpp:54]: Added, thanks @SunjunKim

**Review by SunjunKim — APPROVED** [review]
Checked the functionality and it's working fine.
